### ISSUE TITLE
Question: Disk Types

### ISSUE THREAD

**Issue by meteorfox** [GoogleCloudPlatform/PerfKitBenchmarker#284]
Now that the DISK_TYPE dictionary has been refactored to each provider's disk implementation, what's the best way to represent local/remote disks that can both be spinning disks and/or SSD disks?

Under disk.py I can only see these types defined

```python
STANDARD = 'standard'
REMOTE_SSD = 'remote_ssd'
PIOPS = 'piops'  # Provisioned IOPS (SSD) in AWS
LOCAL = 'local'
```
In Rackspace's case, depending on the flavor (a.k.a machine-type/instance-type), disks could be local or remote, and they can be either spinning disks or SSDs.

Perhaps introduce REMOTE_HDD constant? What do you think?

**Comment by cmccoy** [comment]
`REMOTE_HDD` is equivalent to `STANDARD` currently. 

To handle all of these cases, we'd need to:

* Rename `STANDARD` to `REMOTE_HDD`.
* Rename `LOCAL` to `LOCAL_SSD`.
* Introduce `LOCAL_HDD`.

@ehankland, what do you think?

**Comment by meteorfox** [comment]
@cmccoy I thought `STANDARD` meant whatever is the default when booting an instance.

Let me give some more context, to make sure you have all the information.

In Rackspace, the default type of storage depends on the flavor, for example:
*  `compute1-*` and `memory1-*` by default uses what we are calling `REMOTE_SSD`
  * it could also be `REMOTE_HDD`
* `general1-*` flavors use `LOCAL_HDD`,
* `io-*` flavors use `LOCAL_SSD`,
*  `onmetal-*` use something like `LOCAL_SSD` but actually they are NVMe SSDs. (These are fast!!)

Disk Volumes can only be attached to certain flavors, and can be either SSD or HDD.

**Comment by ehankland** [comment]
@cmccoy 
I think being explicit about remote_hdd is probably a good idea since it's obviously caused some confusion. Splitting up local into local hdd and local ssd would be a bit more difficult. We'd either have to keep track of which instances have local ssd/hdd or we'd have to do the detection after the instances have already booted (this also assumes that all disks correctly report whether they are rotational which may not be universally true). My assumption would be that if a user specified "local_hdd", but the instance only had local ssd (or vice versa) we'd raise an exception and refuse to run. If we did the detection after boot this could be fairly annoying, since all the resources would be created and then torn down. That still leaves the keeping track of machine type/disk type option. I'm open to that if everyone agrees that we should do it, but we basically already encode the disk type in sample metadata since we record the cloud and machine type, so the only real advantage we'd get out of splitting them (at least that I can see) is refusing to run. Another option is simply adding more explicit metadata about the disk type if that's what people wanted out of the split. If anyone thinks there are compelling reasons to split local, please continue the discussion in this thread - it would be useful to hear more opinions.

@meteorfox 
I couldn't seem to find info about Rackspace's local hdd offerings - I could only see local ssds on the pricing page: http://www.rackspace.com/cloud/public-pricing (the page says that "general" types have ssds for the system disk and don't have local data disks) - this page seems to agree with that as well: http://docs.rackspace.com/servers/api/v2/cs-devguide/content/server_flavors.html

**Comment by voellm** [comment]
Good idea.  In fact Eric and I debated the naming of STANDARD to be
REMOTE_HDD or not.  We should just do it.

In terms of LOCAL_SSD vs LOCAL_HDD we also discussed this.  This would
require a lot more knowledge of VM's under test than PKB currently has.  We
should discuss this point more.


On Thu, May 28, 2015 at 2:03 PM, Carlos Torres <notifications@github.com>
wrote:

> Now that the DISK_TYPE dictionary has been refactored to each provider's
> disk implementation, what's the best way to represent local/remote disks
> that can both be spinning disks and/or SSD disks?
>
> Under disk.py I can only see these types defined
>
> STANDARD = 'standard'
> REMOTE_SSD = 'remote_ssd'
> PIOPS = 'piops'  # Provisioned IOPS (SSD) in AWS
> LOCAL = 'local'
>
> In Rackspace's case, depending on the flavor (a.k.a
> machine-type/instance-type), disks could be local or remote, and they can
> be either spinning disks or SSDs.
>
> Perhaps introduce REMOTE_HDD constant? What do you think?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/PerfKitBenchmarker/issues/284>.
>



-- 
Anthony F. Voellm (aka Tony)
Google Voice:  (650) 516-7382
<https://www.google.com/voice/b/0?pli=1#phones>
Blog: http://perfguy.blogspot.com/

**Comment by meteorfox** [comment]
@ehankland whoops, that's my mistake. You are right, I confused the `general` flavors with our old `standard` flavors. They renamed all of them recently, and I still get confused. :)

Now these are the correct ones:
* Local SSDs
  * `general1-*` (Multiple RAID10 SSDs)
  * `io1-*`  (Multiple RAID10 SSDs)
  * `onmetal-(compute|memory)1` (32GB SATADOM)
  * `onmetal-io1`  (32GB SATADOM & 2 x 1.6 TB PCIe Flash, bare-metal)
* Local HDDs
  * `standard`  (1x RAID10 SATA HDDs) - **Deprecated**
* Remote volume (SSD and/or HDD) can be attached to any of these flavors
  * `compute1-*`
  * `memory1-*`
  * `general1-*`
  * `io1-*`

**Comment by meteorfox** [comment]
@cmccoy I'm trying to understand how the ScratchDisk abstraction works.

By default some of our flavors only use local disks (e.g. `general1-*`, `io1-*`, `onmetal-*`, `standard`) with the option to attach remote volumes (except for `onmetal`, and `standard`). So whenever the user executes `pkb.py` with any of those flavors, I want the default to use those local disks provided by the flavor and skip the remote storage volume creation. If there are more than extra 1 local disk available, then it would stripe them together in RAID0. If the flavor does not have extra local disks, then perfkit would just use that (the system disk).

Optionally, I want to let the user specify any number of remote storage volumes (up to 14, and if supported by the flavor), that he/she could attach, then also stripe them together (only remote volumes would be striped, if this is specified), in this case, perfkit would use these volumes for the benchmarks.

The part I don't understand completely is, how is the system disk (e.g. where the OS is running) considered? Is that considered a local scratch disk? I guess my confusion comes from not understanding the whole design to handle disk types, from the CLI flags available to the user, to the actions emitted by perfkit.

It'd be awesome if there we some kind of guide explaining these abstractions, and show examples on how one would go about implementing for a provider. :)

Thanks in advance!

**Comment by ehankland** [comment]
Up until now we haven't let users make use of the system disk as a "scratch disk", but as we are adding more cloud providers, we are running into more clouds where it makes sense to allow that ( @klausw encountered this when adding digital ocean support).  So it's pretty likely that we'll make changes in order to support it.

There are really only 3 flags which govern which disks get used:

flags.DEFINE_enum(
    'scratch_disk_type', disk.STANDARD,
    [disk.STANDARD, disk.REMOTE_SSD, disk.PIOPS, disk.LOCAL],
    'Type for all scratch disks. The default is standard')
flags.DEFINE_integer('num_striped_disks', 1,
                     'The number of disks to stripe together to form one '
                     '"logical" scratch disk. This defaults to 1 '
                     '(except with local disks), which means no striping. '
                     'When using local disks, they default to striping '
                     'all disks together.',
                     lower_bound=1)
flags.DEFINE_integer('scratch_disk_size', 500, 'Size, in gb, for all scratch '
                     'disks, default is 500')

The scratch_disk_type flag determines whether we try to use  remote magnetic ("standard"), remote ssd, or local disks (piops is provisioned iops and is basically AWS specific - at least with the clouds that have been added so far) .
The num_striped_disks flag determines how many disks of the given type you want to stripe together and treat as a single device. A value of 1 (the default) means no striping - a value of 14 would mean creating 14 disks, striping them together and using that as the "scratch disk" for the benchmark.
The scratch_disk_size flag determines how large each disk should be if that can be controlled (it can't be in the local scratch disk case).

A virtual machine's CreateScratchDisk method takes a disk.BaseDiskSpec (which has been filled out according to those flags) and does the work of creating, attaching, striping, formatting, and mounting the scratch disk. There's an optional helper method that does most of this work for you (virtual_machine.BaseVirtualMachine._CreateScratchDiskFromDisks). Assuming you create a disk.BaseDisk subclass for Rackspace disks, and implement the CreateScratchDisk method, that should be just about everything you need. 

There are a few optional steps:

* If necessary, you can implement SetupLocalDisks() - if there's anything you want/need to do (such as unmounting an automounted drive or running a script) in order to prepare the local disks, you can do it here. This will only be executed if --scratch_disk_type=local is specified.

* There is a virtual_machine.BaseVirtualMachine attribute "max_local_disks":
    max_local_disks: The number of local disks on the VM that can be used as
      scratch disks or that can be striped together.
It defaults to 0, but if you set it to the number of local drives on the system, PKB will automatically stripe all of them together when running with --scratch_disk_type=local (unless you specify --num_striped_disks which will override that).

*  Implement GetLocalDisks() which will provide a list of all the raw device paths on the vm. Only one benchmark makes use of this currently (aerospike) and only when using --scratch_disk_type=local. This will probably be removed at some point since it doesn't fit with the way the rest of the benchmarks make use of disks.

If you need more info or if anything I said wasn't clear, please let me know.

**Comment by klausw** [comment]
Based on my current experiments, I think it would help to split the "scratch disk" abstraction into three separate cases:

* **single data directory:** a benchmark needs to store a substantial amount of data, for example *fio* or *bonnie++*, but just uses the mount point it gets from `vm.GetScratchDir()` as a plain directory. This could be striped for performance, but that isn't strictly necessary to execute the benchmark, and could use a plain directory within the base image. It would help if the benchmark would specify how much storage it actually needs - assuming the current default of 500G for scratch disks as a minimum seems excessive.

* **multiple data directories:** a benchmark uses multiple directories that are assumed to be on separate disks, for example *copy_throughput* which uses `vm.GetScratchDir(N)`. This could technically work by creating multiple directories, but that's probably not a good idea if the results would not be comparable to using actual independent disks.

* **disk devices**: a benchmark specifically needs one or more disk devices (not directories) to use internally, i.e. for raw data storage or custom filesystems. This includes *aerospike*. 

I'm currently using a hack which just creates and returns a plain directory for vm.GetScratchDir(), this seems initially promising and gets a few scratch disk based benchmarks working. But I think it would be cleaner to explicitly distinguish the cases through the API.

**Comment by meteorfox** [comment]
@ehankland This is awesome. Ok, it's starting to make sense. I'll let you know if I other questions come into mind. But as you mentioned, Rackspace is also one of those providers that gives you extra local drives, separate from the system disk. I'll probably end up doing the same hack @klausw did, to support using a directory in the system disk as a ScratchDir.

@klausw I think you are onto something.  Some benchmarks need a mounted path in the file system (choice of file system matters), while others might take the path to the block device directly. In the case of benchmarks that only take a mounted path, then it will depend how you chose to setup the block devices behind that mount (e.g. stripe, mirror, writethrough, writeback, etc). Perhaps we should push the task of setting up the file system mounts, and the block devices to the benchmark, then the benchmark decides, where and how to create the ScratchDir for the benchmark to run. But probably each disk will need some metadata describing the nature of the block devices (e.g. system disk, remote, local, ssd, hdd), if they are mounted or not, where are they mounted etc. All of this seems to be the motivation behind `BaseDisk`, as far as I can tell.

So, maybe, drop `GetScratchDir()` from the `BaseVirtualMachine`, and delegate the responsibility of setting up the storage to the benchmarks, by adding a new method to `BaseVirtualMachine` that returns all the available block devices?

**Comment by voellm** [comment]
While we are giving this should we also include ROOT_DISK?

**Comment by klausw** [comment]
@meteorfox - Benchmarks should have the option of setting up their own storage (aerospike already does this), but I think it would still be useful to let a benchmark say "give me a data directory" where the details of this are taken care of by the specific VM implementation combined with flag settings. So I'd want to keep something like GetScratchDir(), just being a bit more explicit about what the benchmarks wants to do with it. For example, the disk spec could say "minimum 20G space, lots of I/O". On GCP this would pick a 500G scratch disk to ensure fast I/O, and on DigitalOcean it would just give you a directory within the image's base partition.

Changing the VM API to only exports block devices would be annoying for DigitalOcean since there's only one device that's already in use for the OS filesystem, and according to [this post](http://digitalocean.uservoice.com/forums/136585-digital-ocean/suggestions/2915575-allow-client-resize-the-dev-hda) it's not resizable.

**Comment by meteorfox** [comment]
@klausw Great point. I agree, seen from that perspective I can see the value of something like GetScratchDir() taking care of the details.

**Comment by meteorfox** [comment]
@ehankland Now that I wrote the scratch_disk_type and machine_type handling, I think I've settled on a way to manage the different options.

Right now:
`STANDARD`, `REMOTE_SSD`, `LOCAL`, and `PIOPS` are the support scratch_disk_types.

Except for `PIOPS`, I tried to map each scratch_disk_type to a Rackspace supported storage, for those that make sense.

Here's how I'm doing the mapping, this follows what I think is 'best fit' as compared to other providers:

![image](https://cloud.githubusercontent.com/assets/232300/7993563/89afb454-0acc-11e5-9637-1966c924edd7.png)


What I want:

I think another approach is to use `STANDARD` to represent the machine_type's strengths in terms of storage. Introduce `REMOTE_HDD`, to specify non-SSD remote volumes, and introduce `BOOT_DISK` to specify that a path in the system disk should be use.

They idea of using `STANDARD` is that if a user doesn't specify a scratch_disk_type, then it would default to what we think is the disk type in terms of performance.

Here's how the compatibility matrix would look for Rackspace with this approach:

![image](https://cloud.githubusercontent.com/assets/232300/7993530/5334329c-0acc-11e5-80d5-48b9c0871880.png)

What do you think?

**Comment by ehankland** [comment]
I talked this over with @voellm and @klausw and we liked your suggestion. From the discussion, I think we also want to rename the new `STANDARD` type to `RECOMMENDED` and `LOCAL` to `LOCAL_EPHEMERAL` (in addition to the name changes you proposed)  to be very explicit about the types so there isn't any confusion.

**Comment by voellm** [comment]
Looks like we all agreed on what to do here.  Ok to close?

**Comment by meteorfox** [comment]
I'm not sure if we should close yet, should we keep track of the changes here? or should we open a new issue to track those changes?

**Comment by voellm** [comment]
Carlos covered this at the recent community meeting here - https://docs.google.com/presentation/d/1XLx3pnwAzI0XHohWxcNQbxnN91jcJXYSAPd6xFfD6Hw/edit#slide=id.g64899ba87_1_9

@noahl is going to work on the next set of changes with Carlos.

**Comment by noahl** [comment]
Hi @meteorfox 

I've been looking through the code and at your slides from the community meeting. There are several different issues brought up here - what the interface for getting scratch disk / directories should be, what should happen when a provider doesn't have a certain type of disk - but I don't think we should let the other issues distract from the renaming project you're proposing. Changing the names of disk categories is something we can do right now, with relatively little work, and it will have a positive impact on PKB. I think we should start with that and hopefully get it merged using the existing infrastructure, while also thinking about what the disk infrastructure should look like.

And speaking of the disk categories, there's an issue I'm struggling with on disks like PD or EBS. In a situation where you choose the size of your drive, and pay more for more performance, what is the "default" size? For best performance, I'd have to recommend spending 100% of your budget on a gigantic PD drive with a huge number of IOPS -- but that's not the best choice for anyone. It's also not clear how to compare the drives. We could try to match drives of the same size, or same maximum IOPS, or same cost. I think same cost is the best choice, but I'm not sure. @voellm?

**Comment by meteorfox** [comment]
@noahl 

I agree with you, I think we can focus on the renaming pieces first, and maybe add the `remote_hdd` disk type as a first step.

After that, I suggest to pay our attention on deciding what `standard` should be replaced with.  I agree that when defining the default disk type, it should not be what is the best disk. Maybe instead we could make it be whatever is the common denominator of all providers, say, using the disk with the root partition, one benefit, is that this will probably be the cheapest option too, and follows the same 'philosophy' as the default flavor. If we go with this approach, then there wouldn't be a need for the `root_disk` type.

Alternatively, the default disk could be whatever disk that comes by default with the machine type. Example, in Rackspace, a machine type of the `general1` class comes with 1 disk, the root partition, even though you can attach remote volumes to the instance, the default for this one would be `root_disk`. Now, for the `onmetal-io1` machine type, the default could be `local_ephemeral` since this machine type already comes with the extra PCI flash drive, and the user is not incurring on an extra cost as opposed to the remote volumes.

That being said, I know this will bring up the issue of how do we handle the case for when benchmarks use the raw blk device, instead of a file. Possibly we could make the `BaseDisk` object include a new method or property that indicates the disk cannot be used as a raw device, then have the benchmarks explicitly define that they need a raw device, and whenever the scratch disk does not satisfy such requirement then PKB could just throw a `NotSupportedConfiguration` error.

What are your thoughts? @voellm

**Comment by noahl** [comment]
Hi @meteorfox,

I just wanted to let you know that we haven't dropped this. We're talking about the taxonomy, and we'll get back to you with our thoughts on the best way to describe disks.

Thanks for your patience,
Noah

**Comment by noahl** [comment]
We at Google have been talking about this more, and we have a proposed design. Anyone who is interested, please comment on this doc: https://docs.google.com/document/d/1ruXqIEg_QAGAD9kpH5rskFVltLtM6gZLqAR67UQg9Es/edit?usp=sharing

**Comment by noahl** [comment]
The disk types refactor was merged in #599. Closing this for now. Please reopen if there is more to do.
